### PR TITLE
Spring Integration javaagent instrumentation

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**

### CODE REVIEW

File: instrumentation/spring/spring-integration-4.1/javaagent/spring-integration-4.1-javaagent.gradle
Comment (mateuszrzeszutek) [inline instrumentation/spring/spring-integration-4.1/javaagent/spring-integration-4.1-javaagent.gradle:7]: This is a consequence of always using the context extracted from the incoming message; we completely ignore the "local" current context, but that's fine in that case.

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (mateuszrzeszutek) [inline instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:53]: This test runs both RabbitMQ and spring-integration instrumentation - the interesting thing here is that the "consumer" span created by spring-integration instrumentation is the child of the actual PRODUCER span (and sibling of CONSUMER) - that's another consequence of choosing the context extracted from the message.

Removing the integration span would be even worse here, since now the parent of whatever comes next would be the PRODUCER span, not the CONSUMER one.

File: instrumentation/spring/spring-integration-4.1/testing/src/main/groovy/AbstractComplexPropagationTest.groovy
Comment (mateuszrzeszutek) [inline instrumentation/spring/spring-integration-4.1/testing/src/main/groovy/AbstractComplexPropagationTest.groovy:30]: FYI: these tests haven't really changed, I just moved them to the `testing` module and made them abstract. I don't understand why git chose to show them as completely new if 95% of content is the same 🤷

File: gradle/instrumentation.gradle
Comment (iNikem) [inline gradle/instrumentation.gradle:105]: Just inline this boolean into the value of system property?

File: instrumentation/spring/spring-integration-4.1/javaagent/spring-integration-4.1-javaagent.gradle
Comment (iNikem) [reply instrumentation/spring/spring-integration-4.1/javaagent/spring-integration-4.1-javaagent.gradle:7]: Hmm... our `debugContextLeakIfEnabled` should not be called at all in case of `CONSUMER` spans. This is not spring integration specific problem.

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (iNikem) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:53]: Which of the spans above is created by spring-integration? 4 or 5?

File: instrumentation/spring/spring-integration-4.1/javaagent/spring-integration-4.1-javaagent.gradle
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-integration-4.1/javaagent/spring-integration-4.1-javaagent.gradle:7]: It's called for all spans, span kind does not matter.

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:53]: Number 5; number 4 (the CONSUMER one) is created by rabbit instrumentation.

File: instrumentation/spring/spring-integration-4.1/javaagent/spring-integration-4.1-javaagent.gradle
Comment (iNikem) [reply instrumentation/spring/spring-integration-4.1/javaagent/spring-integration-4.1-javaagent.gradle:7]: Yes, but the logic of not failing the test should be the same for all `CONSUMER` spans.

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (iNikem) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:53]: I don't like it... I am not sure I want to block this PR, but at some time in near future we have to review span modelling here...

Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:53]: I have mixed feeling about it too - this situation was the reason why my first PR had initially used links for spans extracted from incoming messages.

There's one more idea that comes to my mind, a bit more complex behavior:
* if there's no "local" current span, use the span extracted from the message as the parent;
* otherwise use the "local" span as the parent and link the span context extracted from the message (if it's a different span than parent)

Comment (iNikem) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:53]: We can discuss this tomorrow face to face, if you wish

**Review by iNikem — APPROVED** [review]

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:53]: Yep, definitely! 👍

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (trask) [inline instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:44]: can you add attribute verification? I think that will help me understand the span structure a bit better

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (mateuszrzeszutek) [inline instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:63]: Two CONSUMER spans here are a necessity: turns out spring-cloud-stream for rabbitmq uses a listener that puts all incoming messages into a `BlockingQueue` immediately after receiving them; there's no way to propagate context through that... (BTW, this explains why `@RabbitListener` did not work for some people)

So when spring-integration instrumentation kicks in, it sees no CONSUMER span, so it creates a new one - a child of the remote PRODUCER span.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java
Comment (mateuszrzeszutek) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java:94]: SERVER and CONSUMER spans are now both treated as "local-root" type of spans; by default, you can't have both of them in a single process, in a single trace, since they both represent the receiving side.
This is a pretty significant change, please tell me WDYT.

**Review by iNikem — APPROVED** [review]

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (trask) [inline instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:86]:
```suggestion
          name "testConsumer.input process"
```

**Review by trask — APPROVED** [review]

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (trask) [inline instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:57]: I think messaging semantic conventions suggest this should be just `testTopic send`

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (trask) [inline instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:45]: can you comment in the test which spans are from spring integration instrumentation and which are from rabbitmq instrumentation?

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:57]: That span comes from rabbit instrumentation, do you mind if I create an issue for this?

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:45]: Good point, will do.

File: instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy
Comment (iNikem) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:59]: It seems that we should review rabbitmq integration's usage of semantic convention anyway. There are some other strange things here

Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-integration-4.1/javaagent/src/test/groovy/SpringIntegrationAndRabbitTest.groovy:59]: https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/3337

**Pull request merged**
